superbench/benchmarks/model_benchmarks/pytorch_base.py

```diff
@@ -35,10 +44,243 @@ def __init__(self, name, parameters=''):
+            logger.info('SDP kernel not available')
+            # Older PyTorch versions may not expose sdp_kernel; ignore in that case
+            pass
+
+    def _assign_model_run_metadata(self, precision, extra_keys=None):
+        """Assign model_run_metadata for determinism fingerprinting/logging.
+
+        Args:
+            precision: Model precision (can be enum or string).
+            extra_keys: List of additional argument keys to include in metadata.
+            self._args: Benchmark arguments containing model configuration.
+
+        Returns:
+            None
+        """
+        # Common metadata keys
+        metadata = {
+            'model_name': self._name,
+            'precision': (precision.value if hasattr(precision, 'value') else str(precision)),
+            'seed': getattr(self._args, 'deterministic_seed', None),
+            'deterministic_seed': getattr(self._args, 'deterministic_seed', None),
+            'batch_size': getattr(self._args, 'batch_size', None),
+            'seq_len': getattr(self._args, 'seq_len', None),
+            'num_steps': getattr(self._args, 'num_steps', None),
+            'check_frequency': getattr(self._args, 'check_frequency', None),
+            'num_classes': getattr(self._args, 'num_classes', None),
+        }
+        # Add any extra keys present in args (for model-specific fields)
+        keys = [
+            'hidden_size',
+            'num_hidden_layers',
+            'num_attention_heads',
+            'intermediate_size',
+            'input_size',
+            'num_layers',
+            'bidirectional',
+        ]
+        if extra_keys:
+            keys += extra_keys
+        for key in keys:
+            metadata[key] = getattr(self._args, key, None)
+        self._model_run_metadata = metadata
+        return None
+
+    def record_determinism_fingerprint(self, curr_step, loss, logits, periodic, check_frequency):
+        """Centralized logic for recording per-step loss and periodic fingerprints for deterministic runs.
+
+        Args:
+            curr_step (int): Current training step.
+            loss (torch.Tensor or float): Loss value for this step.
+            logits (torch.Tensor or float): Logits output for this step (sample 0).
+            periodic (dict): Dictionary to store periodic fingerprints ('loss', 'act_mean', 'step').
+            check_frequency (int): Frequency for fingerprint logging.
+        """
+        # Record per-step loss for determinism checks (for full history)
+        try:
+            v = float(loss.detach().item()) if hasattr(loss, 'detach') else float(loss)
+        except Exception:
+            v = None
+        # Periodic fingerprint logging
+        if getattr(self._args, 'deterministic', False) and (curr_step % check_frequency == 0):
+            # 1) Loss fingerprint (only at fingerprinting frequency)
+            try:
+                if 'loss' in periodic and v is not None:
+                    periodic['loss'].append(v)
```

Comment:
Is it fine that the length of `periodic['loss']` may be different with the length of `periodic['step']`?